internal/client/consensus/grandpa/communication_gossip.go

```diff
@@ -0,0 +1,1708 @@
+	switch i.pendingCatchUp.(type) {
+	case pendingCatchUpProcessing:
+		i.pendingCatchUp = pendingCatchUpNone{}
+	default:
+		logger.Debugf("Noted processed catch up message when state was: %v", i.pendingCatchUp)
+	}
+}
+
+func (i *inner[H, N, Hasher]) handleCatchUpRequest(
+	who peerid.PeerID, request catchUpRequestMessage, setState *SharedVoterSetState[H, N]) (gossipMessage, action) {
+	if i.localView == nil {
+		return nil, actionDiscard[H]{ReputationChange: misbehaviorOutOfScopeMessage{}.cost()}
+	}
+
+	if request.SetID != i.localView.setID {
+		// NOTE: When we're close to a set change there is potentially a race where the peer sent us the request before
+		// it observed that we had transitioned to a new set. In this case we charge a lower cost.
+		if saturating.Add(request.SetID, 1) == i.localView.setID &&
+			saturating.Sub(uint64(i.localView.round), catchUpThreshold) == 0 {
+			return nil, actionDiscard[H]{ReputationChange: honestOutOfScopeCatchUp}
+		}
+
+		return nil, actionDiscard[H]{ReputationChange: misbehaviorOutOfScopeMessage{}.cost()}
+	}
+
+	peer := i.peers.peer(who)
+	if peer == nil {
+		return nil, actionDiscard[H]{ReputationChange: misbehaviorOutOfScopeMessage{}.cost()}
+	} else {
+		if peer.view.round >= request.Round {
+			return nil, actionDiscard[H]{ReputationChange: misbehaviorOutOfScopeMessage{}.cost()}
+		}
+	}
+
+	setState.innerMtx.RLock()
+	lastCompletedRound := setState.inner.lastCompletedRound()
+	setState.innerMtx.RUnlock()
+	if Round(lastCompletedRound.Number) < request.Round {
+		return nil, actionDiscard[H]{ReputationChange: misbehaviorOutOfScopeMessage{}.cost()}
+	}
+
+	logger.Tracef("Replying to catch-up request for round %d from %s with round %d",
+		request.Round, who, lastCompletedRound.Number)
+
+	prevotes := make([]grandpa.SignedPrevote[H, N, primitives.AuthoritySignature, primitives.AuthorityID], 0)
+	precommits := make([]grandpa.SignedPrecommit[H, N, primitives.AuthoritySignature, primitives.AuthorityID], 0)
+
+	// NOTE: the set of votes stored in `LastCompletedRound` is a minimal set of votes, i.e. at most one equivocation
+	// is stored per voter. The code below assumes this invariant is maintained when creating the catch up reply since
+	// peers won't accept catch-up messages that have too many equivocations (we exceed the fault-tolerance bound).
+	for _, vote := range lastCompletedRound.Votes {
+		switch message := vote.Message.(type) {
+		case grandpa.Prevote[H, N]:
+			prevotes = append(prevotes, grandpa.SignedPrevote[
+				H, N, primitives.AuthoritySignature, primitives.AuthorityID,
+			]{
+				Prevote:   message,
+				Signature: vote.Signature,
+				ID:        vote.ID,
+			})
+		case grandpa.Precommit[H, N]:
+			precommits = append(precommits, grandpa.SignedPrecommit[
+				H, N, primitives.AuthoritySignature, primitives.AuthorityID,
+			]{
+				Precommit: message,
+				Signature: vote.Signature,
+				ID:        vote.ID,
+			})
+		default:
```

Comment:
if we ever enter in the default, should it not be a `panic("unreacheble")`?

Reply:
No, we are only appending the prevotes and precommits.